internal/transport/http2_server.go

```diff
@@ -375,16 +375,39 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 		}
 	}
 
-	if !isGRPC || headerError {
+	if frame.StreamEnded() && !isGRPC || headerError {
```

Comment:
> In both cases, we would RST_STREAM with NO_ERROR after returning the trailers unless the client has sent an END_STREAM in the initial headers (unexpected for gRPC, but certainly possible in HTTP/2).

Reply:
I did not fix this - I'll leave this for another PR unless the logic directly conflicts with one of the other two changes.